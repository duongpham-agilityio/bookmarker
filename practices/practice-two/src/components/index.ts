export {
  default as Notification,
  type NotificationProps,
} from './Notification';

export { default as Form, type FormProps } from './Form';
export { default as Popup, type PopupProps } from './Popup';
export { default as Card, type CardProps } from './Card';
export { default as CardSkeleton } from './Card/Skeleton';
