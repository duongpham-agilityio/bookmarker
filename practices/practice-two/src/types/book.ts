export type Book = {
  name: string;
  author: string;
  imageURL: string;
  description: string;
  publishDate: number;
  createdAt: number;
  updatedAt: number;
  deletedAt: number | null;
};
