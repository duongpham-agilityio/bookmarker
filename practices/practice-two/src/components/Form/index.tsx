import { MouseEvent, memo } from 'react';

// Components
import { Button, Heading, Input, TextArea } from 'components/commons';

// Types
import { Book } from 'types';

// Styles
import styles from 'components/Form/index.module.css';

// Assets
import UploadIcon from 'assets/icons/upload.svg';
import { useForm } from 'hooks';

export type FormProps = {
  value: Omit<Book, 'deletedAt' | 'createdAt' | 'updatedAt'>;
  title?: string;
  className?: string;
  onClose?: (_event: MouseEvent) => void;
  handleSubmit?: () => void;
};

const Form = (props: FormProps) => {
  const {
    value: data,
    title = 'Create a new book marker',
    className = '',
    onClose,
    handleSubmit,
  } = props;
  const {
    value: { author, description, imageURL, name, publishDate },
    refImage,
    onChange,
    onSubmit,
  } = useForm(data, handleSubmit);

  return (
    <section className={styles.overlay}>
      <form
        className={`${styles.form} ${className}`}
        action="#"
        method="POST"
        onSubmit={onSubmit}
      >
        <Heading label={title} className={styles.heading} />

        <div className={styles.content}>
          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Book name
              </label>
              <Input
                value={name}
                placeholder="Book name"
                name="name"
                id="bookName"
                onChange={onChange}
              />
            </div>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="author">
                Author
              </label>
              <Input
                value={author}
                placeholder="Author"
                name="author"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="publishedDate">
                Published date
              </label>
              <Input
                type="date"
                value={publishDate}
                name="publishedDate"
                onChange={onChange}
              />
            </div>
            <div className={`${styles.formField} ${styles.upload}`}>
              <div className={styles.uploadItem}>
                <label className={styles.label} htmlFor="imageURL">
                  Upload
                </label>
                {!refImage.current?.files?.length ? (
                  <>
                    {/* <Input
                      className={styles.input}
                      type="file"
                      name="imageURL"
                      id="imageURL"
                      accept=".png,.jpeg"
                      onChange={onChange}
                    /> */}
                    <input
                      ref={refImage}
                      hidden
                      className={styles.input}
                      type="file"
                      name="imageURL"
                      id="imageURL"
                      accept=".png,.jpeg"
                    />

                    <Button
                      label="Upload"
                      width="w-lg"
                      border="b-lg"
                      variant="primary"
                      leftIcon={UploadIcon}
                      onClick={() => {
                        refImage.current?.click();
                      }}
                    />
                  </>
                ) : (
                  <p>select image</p>
                )}
              </div>
              <div className={styles.uploadItem}>
                {/* <img src={imageURL} alt="image book" className={styles.image} /> */}
                <div className={styles.image}>{imageURL}</div>
              </div>
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Description
              </label>
              <TextArea
                className={styles.text}
                value={description}
                placeholder="Book name"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.action}>
            <Button
              label="Cancel"
              width="w-lg"
              border="b-lg"
              onClick={onClose}
            />
            <Button
              type="submit"
              label="Save"
              variant="primary"
              width="w-lg"
              border="b-lg"
            />
          </div>
        </div>
      </form>
    </section>
  );
};

export default memo(Form);
