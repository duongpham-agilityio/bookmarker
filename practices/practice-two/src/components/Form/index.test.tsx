import { fireEvent, render } from '@testing-library/react';

// Component
import { Form } from 'components';
import { Book } from 'types';

jest.mock('assets/icons/x-letter.svg', () => ({
  default: 'assets/icons/x-letter.svg',
}));

jest.mock('assets/icons/upload.svg', () => ({
  default: 'assets/icons/upload.svg',
}));

const value: Omit<Book, 'createdAt' | 'deletedAt' | 'updatedAt'> = {
  author: 'John',
  name: 'HTML/CSS',
  description: 'HTML/CSS with styles',
  imageURL: '',
  publishDate: new Date().getTime(),
};

describe('FormComponent', () => {
  it('match snapshot', () => {
    const { asFragment } = render(<Form value={value} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('render', () => {
    const { getByText } = render(<Form value={value} />);

    expect(getByText(/Save/i)).toBeInTheDocument();
  });

  it('action', () => {
    const mockHandleSubmit = jest.fn();
    const mockOnClose = jest.fn();

    const { getByText } = render(
      <Form
        value={value}
        handleSubmit={mockHandleSubmit}
        onClose={mockOnClose}
      />
    );
    const saveBtn = getByText(/Save/i);
    const cancelBtn = getByText(/Cancel/i);

    // click button submit
    fireEvent.click(saveBtn);

    // click button cancel
    fireEvent.click(cancelBtn);

    expect(mockHandleSubmit).toHaveBeenCalled();
    expect(mockOnClose).toHaveBeenCalled();
  });
});
