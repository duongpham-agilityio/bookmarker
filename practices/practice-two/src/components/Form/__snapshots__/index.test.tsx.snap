// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormComponent match snapshot 1`] = `
<DocumentFragment>
  <section>
    <form
      action="#"
      class="undefined "
      method="POST"
    >
      <h2
        class="undefined undefined "
      >
        Create a new book marker
      </h2>
      <div>
        <div>
          <div>
            <label
              for="bookName"
            >
              Book name
            </label>
            <div
              class="undefined "
            >
              <input
                id="bookName"
                name="name"
                placeholder="Book name"
                type="text"
                value="HTML/CSS"
              />
            </div>
          </div>
          <div>
            <label
              for="author"
            >
              Author
            </label>
            <div
              class="undefined "
            >
              <input
                name="author"
                placeholder="Author"
                type="text"
                value="John"
              />
            </div>
          </div>
        </div>
        <div>
          <div>
            <label
              for="publishedDate"
            >
              Published date
            </label>
            <div
              class="undefined "
            >
              <input
                name="publishedDate"
                placeholder="Enter text..."
                type="date"
                value="2023-01-02"
              />
            </div>
          </div>
          <div
            class="undefined undefined"
          >
            <div>
              <label
                for="imageURL"
              >
                Upload
              </label>
              <input
                accept=".png,.jpeg"
                hidden=""
                id="imageURL"
                name="imageURL"
                type="file"
              />
              <button
                class="undefined undefined
  undefined undefined
  undefined "
                type="button"
              >
                <img
                  alt="icon action"
                  src="assets/icons/upload.svg"
                />
                Upload
              </button>
            </div>
            <div>
              <div />
            </div>
          </div>
        </div>
        <div>
          <div>
            <label
              for="bookName"
            >
              Description
            </label>
            <textarea
              class="undefined "
              placeholder="Book name"
            >
              HTML/CSS with styles
            </textarea>
          </div>
        </div>
        <div>
          <button
            class="undefined undefined
  undefined undefined
  undefined "
            type="button"
          >
            Cancel
          </button>
          <button
            class="undefined undefined
  undefined undefined
  undefined "
            type="submit"
          >
            Save
          </button>
        </div>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
